### ISSUE TITLE
Incorrect changes when using ReplaceAroundStep

### ISSUE THREAD

**Issue by floorish** [ProseMirror/prosemirror-changeset#7]
When wrapping some existing change with for example a `blockquote`, the original change is adjusted incorrectly.

## Reproduction

```
<p></p>                               []
[insert ab]                           + [1, 1, 1, 3]
<p>ab</p>                             = [1, 1, 1, 3]
[wrap with blockquote]                + [[0, 0, 0, 1], [4, 4, 5, 6]]
<blockquote><p>ab</p></blockquote>    = [[0, 0, 0, 1], [3, 3, 5, 6], [1, 1, 4, 6]]
```

So the merge of `[[1, 1, 1, 3]]` with  `[[0, 0, 0, 1], [4, 4, 5, 6]]` is incorrect. Expected final change would be:` [[0, 0, 0, 1], [1, 1, 2, 4], [2, 2, 5, 6]]`

To investigate this we tried to simulate a wrap by inserting empty paragraphs around a change. The difference between a ReplaceAroundStep and two wrapping inserts is the resulting StepMap.

ReplaceAroundStep gives a single StepMap: `[0 0 1 4 0 1]` whereas two inserts in a single transaction give two StepMaps: `[0 0 1] + [5 0 1]`. 

This results in different `stepChanges` in the `addSteps` method:
- ReplaceAroundStep StepMap `forEach` gives: `[[0, 0, 0, 1], [4, 4, 5, 6]]`
- Two inserts StepMaps `forEach` gives: `[[0, 0, 0, 1], [5, 5, 5, 6]]`

Because `Change.merge` handles these two inputs differently, the resulting merged Changes is incorrect for the ReplaceAroundStep. It is correct for two inserts transaction.

## Tests
The tests to reproduce this issue are available here: https://github.com/floorish/prosemirror-changeset/commit/2ad1e8f8f9aa6dab29ad3c2b2685e09b563a0339

Tests which describe the non-overlapping merges are available here:
https://github.com/floorish/prosemirror-changeset/commit/ff0837e2748eb03cc7a6f798c6cf872c5fbecfb5

### A
The `forEach` function on StepMap needs adjustment, such that an input of `[0 0 1 4 0 1]` returns `[[0, 0, 0, 1], [5, 5, 5, 6]]`

Implementation of this solution is available here:
https://github.com/floorish/prosemirror-transform/commit/d666b124c955a0c9a77c282e80ec369666b34ae9

### B
StepMap needs adjustment.

- Insert + Insert gives: `[0 0 1] + [5 0 1]`. When remapping pos 4 we'll get: `mapping->map(4) = 5`
- ReplaceAround gives:   `[0 0 1 4 0 1]`. When remapping pos 4 we'll get: `mapping->map(4) = 6`

This seems like inconsistent behaviour, since start-doc and end-doc are in both cases the same.

### C
The `stepChanges` creation in `addSteps` method needs adjustment, such that an input of `[0 0 1 4 0 1]` returns `[[0, 0, 0, 1], [5, 5, 5, 6]]`

Implementation of this solution is available here: https://github.com/floorish/prosemirror-changeset/commit/717aaf6d6cae10142327f8ed48d87d0940727483

### D
The `Change.merge` method currently adjusts the `fromA` and `toA` values, so merging `[0, 0, 0, 2] + [6, 6, 6, 8]` gives `[[0, 0, 0, 2], [4, 4, 6, 8]]`. Maybe something can be changed here such that it keeps the original `fromA` and `toA` values?

These results all differ, which is expected?

```
merge [[1, 1, 1, 3]] + [[0, 0, 0, 2], [4, 4, 6, 8]]
= [[0,0,0,2],[1,1,3,5],[2,2,6,8]]

merge [[1, 1, 1, 3]] + [[0, 0, 0, 2], [6, 6, 6, 8]]
= [[0,0,0,2],[1,1,3,5],[4,4,6,8]]
```

**Comment by floorish** [comment]
@marijnh Perhaps any comments on this issue?

Currently were using fix **C**: https://github.com/floorish/prosemirror-changeset/commit/717aaf6d6cae10142327f8ed48d87d0940727483

If you want I can create a pull request for that as well, but perhaps a solution for **B** is better.

**Comment by marijnh** [comment]
> The forEach function on StepMap needs adjustment, such that an input of [0 0 1 4 0 1] returns [[0, 0, 0, 1], [5, 5, 5, 6]]

That method is defined as producing position referring to the pre- and post-step documents, which would be violated by this, so that doesn't seem like a good direction.

This does indeed look like a bug in StepMap, and your patch https://github.com/floorish/prosemirror-changeset/commit/717aaf6d6cae10142327f8ed48d87d0940727483 looks reasonable—blindly putting changes at the returned positions, as the code was doing before, isn't safe.

**Comment by marijnh** [comment]
I've merged your patch as 54e7214a82abed5 and added the test in 1b665fa5ee5eafc7

**Comment by floorish** [comment]
Cheers, thanks!
